New tutorial for creating a workspace (#359)
ros2/ros2_documentation: source/Tutorials/Workspace/CreatingAWorkspace.rst

.. _ROS2Workspace:

Creating a workspace
====================

**Goal:** Create a workspace and learn how to set up an overlay for development and testing.

**Tutorial level:** Beginner

**Time:** 20 minutes

.. contents:: Contents
   :depth: 2
   :local:

Background
----------

> Before using ROS 2, it’s necessary to source your main ROS 2 installation in the terminal you plan to work in.

Reviewer: I'll note that the name of the tutorial is "Creating a workspace", but there is no definition of a workspace here.  We should probably lead with what a workspace is, then go into the explanation of overlays and underlays.

Author: How about: "A workspace is a directory containing ROS 2 packages"?

Reviewer: > How about: "A workspace is a directory containing ROS 2 packages"?

It's a bit more.  A workspace contains the source for ROS 2 packages, but will also contain the build artifacts, install artifacts, and log artifacts once you build with colcon.

Actually, that previous sentence is probably a good place to start :).

This makes ROS 2’s packages available for you to use in that terminal.
You also have the option of sourcing an “overlay” – a secondary workspace where you can work on packages without interfering with the main ROS 2 installation, or “underlay”.
Your underlay must contain the dependencies of all the packages in your overlay.

> Modifications and developments in your overlay will override packages in the underlay.

Reviewer: ```suggestion
Packages in your overlay will override packages in the underlay.
```

Reviewer: Also, we should probably mention here that you can have "stacked" underlays/overlays (depending your point-of-view).  That is, you can use packages from `/opt/ros/dashing` as underlay 1, then use packages from `~/my_deps` as underlay 2, and then use packages from `~/my_app` as the top-level overlay.  It's somewhat complicated for a beginner (since it is easier to just have the underlay be the system installation and have a single overlay), but it is worth mentioning that it is possible since we mention it off-handedly below.

Prerequisites
-------------

* :ref:`ROS 2 installation <InstallationGuide>`
* :ref:`colcon installation <Colcon>`
* `git installation <https://git-scm.com/book/en/v2/Getting-Started-Installing-Git>`__
* :ref:`turtlesim installation <Turtlesim>`
* Understanding of basic terminal commands (`here’s a guide for linux <http://www.ee.surrey.ac.uk/Teaching/Unix/>`__)
* Text editor of your choice

Tasks
-----

1 Source ROS 2 environment
^^^^^^^^^^^^^^^^^^^^^^^^^^

Your main ROS 2 installation will be your underlay for this tutorial.
(Keep in mind that an underlay does not necessarily have to be the main ROS 2 installation.)

Depending on how you installed ROS 2 (from source or binaries), and which platform you’re on, your exact source command will vary:

.. tabs::

   .. group-tab:: Linux

      .. code-block:: bash

        source /opt/ros/<distro>/setup.bash

      For example, if you installed ROS 2 Dashing:

      .. code-block:: bash

        source /opt/ros/dashing/setup.bash

   .. group-tab:: macOS

      .. code-block:: bash

>         . ~/ros2_install/ros2-osx/setup.bash

Reviewer 2: We wanted to make sure to embed the distro name into the directory for macOS / Windows too, right?

Author: Does this do it #380 ?

In the rest of the beginner developer tutorials, you will create your own packages, but for now you will practice putting a workspace together using existing packages.

The existing packages you will use are from the ``ros_tutorials`` repository (repo).
If you went through the beginner user tutorials, you'll be familiar with ``turtlesim``, one of the packages in this repo.

You can see the repo `on github <https://github.com/ros/ros_tutorials/>`__

Notice the “Branch” drop down list to the left above the directories list.
When you clone this repo, add the ``-b`` argument followed by the branch that corresponds with your ROS 2 distro.

In the ``dev_ws/src`` directory, if your distro is Dashing for example, run the command:

.. code-block:: bash

  git clone https://github.com/ros/ros_tutorials.git -b dashing-devel

Now ``ros_tutorials`` is cloned in your workspace.
Enter ``ls`` in your terminal, and you will see the new ``ros_tutorials`` directory.

To see the packages inside ``ros_tutorials``, enter the command:

.. code-block:: bash

    ls ros_tutorials

Which will list the contents of the repo you just cloned, like so:

.. code-block:: bash

>     roscpp_tutorials  rospy_tutorials  ros_tutorials  turtlesim

Reviewer 2: (Not the greatest example) but we probably should mentioned that three of the four packages are being ignore and `turtlesim` is the only actual ROS 2 package in this repo.

Now you have populated your workspace with sample packages, but it isn’t a fully-functional workspace yet.

> You need to build the workspace first.

Reviewer: Since the next section is "Resolve dependencies", this should probably change to something like:

```
You need to resolve dependencies and build the workspace first.
```

4 Resolve dependencies
^^^^^^^^^^^^^^^^^^^^^^

Before building the workspace, you need to resolve package dependencies.
You may have all the dependencies already, but best practice is to check for dependencies every time you clone.
You wouldn’t want a build to fail after a long wait because of missing dependencies.

From the root of your workspace (``~/dev_ws``):
Run the following command, replacing ``<distro>`` with your distro:

.. code-block:: bash

>   sudo rosdep install -i --from-path src --rosdistro <distro> -y

Reviewer: Do we have a tutorial that talks about `rosdep`?  Otherwise, this is kind of magical.  If we *do* have a tutorial talking about `rosdep`, we should probably have some language saying that you'll learn about it later.

Author: Hmm.. there is a bit more about dependencies once they start creating their own packages, but not much more about `rosdep` specifically. I'll say "later" for now and make a note to add more in-depth notes about `rosdep`. I also think the opening and closing lines of this section provide a sufficient temporary introduction, albeit very vague

Reviewer: I'm fine with punting on rosdep for now.  I've opened https://github.com/ros2/ros2_documentation/issues/381 to track adding a tutorial about rosdep.

For example, if you're using Dashing, you would run:

.. code-block:: bash

  sudo rosdep install -i --from-path src --rosdistro dashing -y

If you already have all your dependencies, the console will return:

.. code-block:: bash

  #All required rosdeps installed successfully

Packages declare their dependencies in the package.xml file (you will learn more about packages in the next tutorial).
This command walks through those declarations and installs the ones that are missing.

5 Build the workspace with colcon
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

From the root of your workspace (``~/dev_ws``), you can now build your packages using the command:

> .. code-block:: bash

Reviewer: Since Windows is different here, can we use the tabs?

  colcon build

If you’re using Windows, you will need to append the ``merge-install`` argument like so:

.. code-block:: bash

  colcon build --merge-install

Windows doesn’t allow long paths, so ``merge-install`` will combine all the paths into the ``install`` directory.

> The cosole will return the following message:

Reviewer: ```suggestion
The console will return the following message:
```

    Other useful arguments for ``colcon build``:

    * ``--packages-up-to`` builds the package you want, plus all its dependencies, but not the whole workspace (saves time)
    * ``--symlink-install`` saves you from having to rebuild every time you tweak python scripts
    * ``--event-handlers console-direct+`` shows console output while building (can otherwise be found in the ``log`` directory)

Once the build is finished, enter ``ls`` in the workspace root (``~/dev_ws``) and you will see that colcon has created new directories:

.. code-block:: bash

  build  install  log  src

The ``install`` directory is where your workspace’s setup files are, which you can use to source your overlay.

6 Source the overlay
^^^^^^^^^^^^^^^^^^^^

Before sourcing the overlay, it is very important that you open a new terminal, separate from the one where you built the workspace.
Sourcing an overlay in the same terminal where you built, or likewise building where an overlay is sourced, may create complex issues.

In the new terminal, source your main ROS 2 environment as the “underlay”, so you can build the overlay “on top of” it:

.. tabs::

   .. group-tab:: Linux

      .. code-block:: bash

>         source /opt/ros/<distro>/setup.bash

Reviewer 2: Sourcing the `local_setup` file would be sufficient (since there are no parent workspace above this) but for consistency it is ok to source the `setup` file. Then the Windows command should do the same though.

   .. group-tab:: macOS

      .. code-block:: bash

        . ~/ros2_install/ros2-osx/setup.bash

   .. group-tab:: Windows

      .. code-block:: bash

        call C:\dev\ros2\local_setup.bat

Go into the root of your workspace:

.. code-block:: bash

  cd dev_ws

> In the root, source your overlay:

Reviewer 2: If the user has already sourced the underlay they only need to source the `local_setup` file of the overlay.

Or they just source the `setup` file of the overlay which implicitly will source the underlay.

Reviewer: Actually, an explanation of the difference between `setup.bash` and `local_setup.bash` would probably be good here.  It's a question I get a lot.

Author: Adding this note:
```
.. note::

  Sourcing the ``local_setup`` of the overlay will only add the packages available in the overlay to your environment.
  ``setup`` sources the overlay as well as the underlay it was created in, allowing you to utilize both workspaces.

  So, sourcing your main ROS 2 installation's ``setup`` and then the ``dev_ws`` overlay's ``local_setup``, like you just did,
  is the same as just sourcing ``dev_ws``'s ``setup``, because that includes the environment of the underlay it was created in.
```
Does that make sense?

You can modify ``turtlesim`` in your overlay by editing the title bar on the turtlesim window.
To do this, locate the ``turtle_frame.cpp`` file in ``~/dev_ws/src/ros_tutorials/turtlesim/src``.
Open ``turtle_frame.cpp`` with your preferred text editor.

On line 52 you will see the function ``setWindowTitle("TurtleSim");``.
Change the value ``”TurtleSim”`` to ``”MyTurtleSim”``, and save the file.

Return to first terminal where you ran ``colcon build`` earlier and run it again.

Return to the second terminal (where the overlay is sourced) and run turtlesim again:

.. code-block:: bash

  ros2 run turtlesim turtlesim_node

You will see the title bar on the turtlesim window now says “MyTurtleSim”.

.. image:: overlay.png

Even though your main ROS 2 environment was sourced in this terminal earlier, the overlay of your ``dev_ws`` environment takes precedence over the contents of the underlay.

If you open another terminal, only source your main ROS 2 distro, and run turtlesim, the title bar will be back to normal (“TurtleSim”).

.. image:: underlay.png

You can see that modifications in the overlay did not actually affect anything in the underlay.

Summary
-------

> In this tutorial, you sourced your main ROS 2 distro install as your underlay, and created an overlay by running ``. install/setup.bash`` in your new workspace.

Reviewer: To be clear, you don't "create" the overlay by sourcing `install/setup.bash`; you bring the overlay into effect.  The creation of the overlay is done by cloning and building the workspace.  So maybe:

```
In this tutorial, you sourced your main ROS 2 distro install as your underlay, and created an overlay by cloning and building packages in a new workspace.
```

.. _ROS2Workspace:

Creating a workspace
====================

**Goal:** Create a workspace and learn how to set up an overlay for development and testing.

**Tutorial level:** Beginner

**Time:** 20 minutes

.. contents:: Contents
   :depth: 2
   :local:

Background
----------

A workspace is a directory containing ROS 2 packages.

> Before using ROS 2, it’s necessary to source your main ROS 2 installation workspace in the terminal you plan to work in.

Reviewer 3: ```suggestion
Before using ROS 2, it’s necessary to source your the ROS 2 installation workspace in the terminal you plan to work in.
```

Reviewer 3: It's not necessary to source the core installation. You can go directly to an overlay to do work.

This makes ROS 2’s packages available for you to use in that terminal.

> You also have the option of sourcing an “overlay” – a secondary workspace where you can work on packages without interfering with the main ROS 2 installation, or “underlay”.

Reviewer 3: ```suggestion
You also have the option of sourcing an “overlay” – a secondary workspace where you can add new packages without interfering with the existing ROS 2 workspace that you're extending, or “underlay”.
```

Reviewer 3: Slightly more generic to not call out the "main" installation since the underlay could be otherwise.

        call C:\dev\ros2\local_setup.bat

Consult the :ref:`installation guide <InstallationGuide>` you followed if these commands don’t work for you.

2 Create a new directory
^^^^^^^^^^^^^^^^^^^^^^^^

Best practice is to create a new directory for every new workspace.
The name doesn’t matter, but it is helpful to have it indicate the purpose of the workspace.
Let’s choose the directory name ``dev_ws``, for “development workspace”:

.. code-block:: bash

  mkdir dev_ws
  mkdir dev_ws/src
  cd dev_ws/src

Another best practice is to put any packages in your workspace into the ``src`` directory.
The above code creates a ``src`` directory inside ``dev_ws`` and then navigates into it.

3 Clone a sample repo
^^^^^^^^^^^^^^^^^^^^^

Ensure you’re still in the ``dev_ws/src`` directory before you clone.

In the rest of the beginner developer tutorials, you will create your own packages, but for now you will practice putting a workspace together using existing packages.

The existing packages you will use are from the ``ros_tutorials`` repository (repo).
If you went through the beginner user tutorials, you'll be familiar with ``turtlesim``, one of the packages in this repo.

> You can see the repo `on github <https://github.com/ros/ros_tutorials/>`__

Reviewer: Nit: add a period at the end of the sentence.

Notice the “Branch” drop down list to the left above the directories list.
When you clone this repo, add the ``-b`` argument followed by the branch that corresponds with your ROS 2 distro.

In the ``dev_ws/src`` directory, if your distro is Dashing for example, run the command:

.. code-block:: bash

  git clone https://github.com/ros/ros_tutorials.git -b dashing-devel

Now ``ros_tutorials`` is cloned in your workspace.

> Enter ``ls`` in your terminal, and you will see the new ``ros_tutorials`` directory.

Reviewer: This bit is Linux/macOS specific.  Since the rest of the tutorial uses the tabbed interface for Linux/macOS/Windows, you probably want to do the same here.  I *think* the command would be `dir ros_tutorials` on Windows.

.. code-block:: bash

  build  install  log  src

The ``install`` directory is where your workspace’s setup files are, which you can use to source your overlay.

6 Source the overlay
^^^^^^^^^^^^^^^^^^^^

Before sourcing the overlay, it is very important that you open a new terminal, separate from the one where you built the workspace.
Sourcing an overlay in the same terminal where you built, or likewise building where an overlay is sourced, may create complex issues.

In the new terminal, source your main ROS 2 environment as the “underlay”, so you can build the overlay “on top of” it:

.. tabs::

   .. group-tab:: Linux

      .. code-block:: bash

        source /opt/ros/<distro>/setup.bash

   .. group-tab:: macOS

      .. code-block:: bash

        . ~/ros2_install/ros2-osx/setup.bash

   .. group-tab:: Windows

      .. code-block:: bash

        call C:\dev\ros2\setup.bat

Go into the root of your workspace:

.. code-block:: bash

  cd dev_ws

In the root, source your overlay:

.. tabs::

  .. group-tab:: Linux

    .. code-block:: bash

      . install/local_setup.bash

  .. group-tab:: macOS

    .. code-block:: bash

      . install/local_setup.bash

>   .. group-tab:: Linux

Reviewer: This should be "Windows".

    .. code-block:: bash

      . install/local_setup.bat

.. note::

  Sourcing the ``local_setup`` of the overlay will only add the packages available in the overlay to your environment.
  ``setup`` sources the overlay as well as the underlay it was created in, allowing you to utilize both workspaces.

  So, sourcing your main ROS 2 installation's ``setup`` and then the ``dev_ws`` overlay's ``local_setup``, like you just did,
  is the same as just sourcing ``dev_ws``'s ``setup``, because that includes the environment of the underlay it was created in.

Now you can run the ``turtlesim`` package from the overlay:

.. code-block:: bash

  ros2 run turtlesim turtlesim_node

But how can you tell that this is the overlay turtlesim running, and not your main installation's turtlesim?

Let’s modify turtlesim in the overlay so you can see the effects:
* You can modify and rebuild packages in the overlay separately from the underlay.

> * The overlay takes precedence over the underlay.

Reviewer: These two bullets don't render properly in the heroku preview.  Do we need a different syntax for RST here?

7 Modify the overlay
^^^^^^^^^^^^^^^^^^^^

You can modify ``turtlesim`` in your overlay by editing the title bar on the turtlesim window.
To do this, locate the ``turtle_frame.cpp`` file in ``~/dev_ws/src/ros_tutorials/turtlesim/src``.
Open ``turtle_frame.cpp`` with your preferred text editor.

On line 52 you will see the function ``setWindowTitle("TurtleSim");``.
Change the value ``”TurtleSim”`` to ``”MyTurtleSim”``, and save the file.

Return to first terminal where you ran ``colcon build`` earlier and run it again.

Return to the second terminal (where the overlay is sourced) and run turtlesim again:

.. code-block:: bash

  ros2 run turtlesim turtlesim_node

You will see the title bar on the turtlesim window now says “MyTurtleSim”.

.. image:: overlay.png

Even though your main ROS 2 environment was sourced in this terminal earlier, the overlay of your ``dev_ws`` environment takes precedence over the contents of the underlay.

> If you open another terminal, only source your main ROS 2 distro, and run turtlesim, the title bar will be back to normal (“TurtleSim”).

Reviewer 3: Giving the commands to run this would be helpful. With a clear reminder to run it in a new terminal to compare.

`source /opt/ros/dashing/setup.bash && ros2 run turtlesim turtlesim_node`

> If you open another terminal, only source your main ROS 2 distro, and run turtlesim, the title bar will be back to normal (“TurtleSim”).

Reviewer 3: ```suggestion
If you open another terminal, only source your underlay, the main ROS 2 distro, and run turtlesim, the title bar will be back to normal (“TurtleSim”).
```
